dissect/target/plugins/apps/browser/firefox.py

```diff
@@ -265,4 +295,324 @@ def downloads(self) -> Iterator[BrowserDownloadRecord]:
+
+    if not decrypted_password_check:
+        log.warning("Encountered unknown algorithm %s while decrypting master key.", algorithm)
+        return b"", ""
+
+    expected_password_check = b"password-check\x02\x02"
+    if decrypted_password_check != b"password-check\x02\x02":
+        log.warning(
+            "Master key decryption failed. Either the master key is protected by a "
+            "primary password which you did not supply, or the primary password you provided was incorrect."
+        )
+        log.debug("Expected %s but got %s", expected_password_check, decrypted_password_check)
+        return b"", ""
+
+    master_key, master_key_cka = query_master_key(key4_file)
+    if master_key == b"":
+        log.warning("Password master key is not defined.")
+        return b"", ""
+
+    if master_key_cka != CKA_ID:
+        log.warning("Password master key cka '%s' is not equal to expected value '%s'.", master_key_cka, CKA_ID)
+        return b"", ""
+
+    decoded_master_key = core.load(master_key)
+    decrypted, algorithm = decrypt_master_key(decoded_master_key, primary_password, global_salt)
+    return decrypted[:24], algorithm
+
+
+def decrypt_field(key: bytes, field: tuple[bytes, bytes, bytes]) -> bytes:
+    if not HAS_CRYPTO:
+        raise ValueError("Missing cryptography dependency")
+
+    cka, iv, ciphertext = field
+
+    if cka != CKA_ID:
+        raise ValueError("Expected cka to equal '%s' but got '%s'", CKA_ID, cka)
+
+    return unpad(DES3.new(key, DES3.MODE_CBC, iv).decrypt(ciphertext), 8)
+
+
+def decrypt(
+    username: str, password: str, key4_file: TargetPath, primary_password: str = ""
+) -> tuple[Optional[str], Optional[str]]:
+    """Decrypt a stored username and password using provided credentials and key4 file.
+
+    Args:
+        username: encoded and encrypted password
+        password encoded and encrypted password
+        key4_file: path to key4.db file
+        primary_password: password to use for decryption routine
+
+    Returns: A tuple of decoded username and password strings.
+
+    Resources:
+        - https://github.com/lclevy/firepwd
+    """
+    if not HAS_CRYPTO:
+        raise ValueError("Missing cryptography dependency")
+
+    try:
+        username = decode_login_data(username)
+        password = decode_login_data(password)
+
+        primary_password_bytes = primary_password.encode()
+        key, algorithm = retrieve_master_key(primary_password_bytes, key4_file)
+
+        if not key or not algorithm:
+            return None, None
+
+        if algorithm == pbeWithSha1AndTripleDES_CBC or algos.EncryptionAlgorithmId.map(algorithm) == "pbes2":
+            username = decrypt_field(key, username)
+            password = decrypt_field(key, password)
+            return username.decode(), password.decode()
+    except ValueError as e:
+        log.error("Failed to decrypt password using keyfile %s and password '%s'", key4_file, primary_password)
```

Comment:
Similar here, maybe nice to replace this log with a raise and do the logging ("error handling") one level higher? Pro-tip, you can include the stacktrace of another exception with `raise ValueError("Blah") from e`